File: packages/zatca/scripts/generate-csr.py
#!/usr/bin/env python3
"""
Generate a ZATCA-compliant ECDSA CSR.

Curve: secp256k1 — confirmed from decoded ZATCA-issued certificates and
working open-source implementations. Documentation says P-256 but the
actual ZATCA gateway uses secp256k1.

Fields go in Subject DN with UTF8String encoding (not SAN extension).
"""

import base64
from cryptography import x509
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec
from cryptography.x509.oid import NameOID
from cryptography.x509.name import _ASN1Type

# ── Configuration ─────────────────────────────────────────────────────────────
VAT_NUMBER        = "310712587200003"
ORG_NAME          = "Tahawwul Telecom and IT Company"
BRANCH_NAME       = "Main Branch"
SERIAL_NUMBER     = "1-Tahawwul|2-EGS1-001|3-MCP"  # SN: 1-name|2-serial|3-solution
CITY              = "Riyadh"
COUNTRY           = "SA"
BUSINESS_TYPE     = "1100"          # 1100=B2B+B2C, 0100=B2B only, 1000=B2C only
BUSINESS_CATEGORY = "Technology"
# ─────────────────────────────────────────────────────────────────────────────

print("Generating ECDSA secp256k1 key pair...")
private_key = ec.generate_private_key(ec.SECP256K1())

subject = x509.Name([
    x509.NameAttribute(NameOID.COUNTRY_NAME,             COUNTRY),
    x509.NameAttribute(NameOID.ORGANIZATION_NAME,        ORG_NAME),
    x509.NameAttribute(NameOID.ORGANIZATIONAL_UNIT_NAME, BRANCH_NAME),
    x509.NameAttribute(NameOID.COMMON_NAME,              ORG_NAME),
    x509.NameAttribute(NameOID.SERIAL_NUMBER,            SERIAL_NUMBER,    _type=_ASN1Type.UTF8String),
    x509.NameAttribute(NameOID.USER_ID,                  VAT_NUMBER,       _type=_ASN1Type.UTF8String),
    x509.NameAttribute(NameOID.TITLE,                    BUSINESS_TYPE,    _type=_ASN1Type.UTF8String),
    x509.NameAttribute(NameOID.STATE_OR_PROVINCE_NAME,   CITY,             _type=_ASN1Type.UTF8String),
    x509.NameAttribute(NameOID.BUSINESS_CATEGORY,        BUSINESS_CATEGORY,_type=_ASN1Type.UTF8String),
])

csr = (
    x509.CertificateSigningRequestBuilder()
    .subject_name(subject)
    .sign(private_key, hashes.SHA256())
)

# Write PEM files
key_pem = private_key.private_bytes(
    encoding=serialization.Encoding.PEM,
    format=serialization.PrivateFormat.TraditionalOpenSSL,
    encryption_algorithm=serialization.NoEncryption(),
)
csr_pem = csr.public_bytes(serialization.Encoding.PEM)

with open("zatca-private-key.pem", "wb") as f:
    f.write(key_pem)
with open("zatca-csr.pem", "wb") as f:
    f.write(csr_pem)

# Base64 DER for the API
csr_der    = csr.public_bytes(serialization.Encoding.DER)
csr_base64 = base64.b64encode(csr_der).decode()

key_der    = private_key.private_bytes(
    encoding=serialization.Encoding.DER,
    format=serialization.PrivateFormat.TraditionalOpenSSL,
    encryption_algorithm=serialization.NoEncryption(),
)
key_base64 = base64.b64encode(key_der).decode()

print("\nFiles written: zatca-private-key.pem, zatca-csr.pem")
print(f"\nSubject: C={COUNTRY}, O={ORG_NAME}, OU={BRANCH_NAME}, CN={ORG_NAME}")
print(f"         SN={SERIAL_NUMBER}")
print(f"         UID={VAT_NUMBER}, title={BUSINESS_TYPE}, businessCategory={BUSINESS_CATEGORY}")
print(f"         registeredAddress={CITY}")

print("\n=== CSR Base64 (paste as \"csr\" in POST /compliance) ===\n")
print(csr_base64)

print("\n=== Private Key Base64 (set as ZATCA_PRIVATE_KEY in .env) ===\n")
print(key_base64)

File: packages/zatca/scripts/generate-csr.mjs
/**
 * Generate a ZATCA-compliant ECDSA CSR using node-forge.
 *
 * ZATCA requires specific OID extensions in the CSR:
 *   2.16.840.1.114412.1.1 = EGS Serial Number
 *   2.16.840.1.114412.1.2 = VAT Registration Number
 *
 * Usage: node generate-csr.mjs
 */

import forge from "node-forge";
import { writeFileSync } from "fs";

// ── Configuration ───────────────────────────────────────────────────────────
const VAT_NUMBER        = "310712587200003";
const ORG_NAME          = "Tahawwul Telecom and IT Company";
const BRANCH_NAME       = "Main Branch";
const SERIAL_NUMBER     = "1-Tahawwul|2-EGS1-001|3-MCP";  // SN field: 1-name|2-serial|3-solution
const CITY              = "Riyadh";
const COUNTRY           = "SA";
const BUSINESS_TYPE     = "1100";   // 1100 = B2B + B2C, 0100 = B2B only, 1000 = B2C only
const BUSINESS_CATEGORY = "Technology";
// ────────────────────────────────────────────────────────────────────────────

// node-forge doesn't natively support ECDSA CSR — use RSA-2048 for sandbox testing.
// Production requires ECDSA P-256 signed with the ZATCA-issued certificate.
console.log("Generating RSA-2048 key pair (ZATCA sandbox accepts RSA)...");
const keys = forge.pki.rsa.generateKeyPair(2048);
const privateKey = keys.privateKey;
const publicKey  = keys.publicKey;

const csr = forge.pki.createCertificationRequest();
csr.publicKey = publicKey;

// ZATCA-required Subject DN fields (section 4.3 of Developer Portal Manual)
// Using OIDs directly where node-forge doesn't know the shortName.
csr.setSubject([
  { name: "countryName",            value: COUNTRY },           // C  2.5.4.6
  { name: "organizationName",       value: ORG_NAME },          // O  2.5.4.10
  { name: "organizationalUnitName", value: BRANCH_NAME },       // OU 2.5.4.11
  { name: "commonName",             value: ORG_NAME },          // CN 2.5.4.3
  // SN (serialNumber 2.5.4.5) = "1-{name}|2-{serial}|3-{solution}"
  { type: "2.5.4.5",               value: SERIAL_NUMBER },
  // UID (userId 0.9.2342.19200300.100.1.1) = VAT number
  { type: "0.9.2342.19200300.100.1.1", value: VAT_NUMBER },
  // title (2.5.4.12) = business type
  { type: "2.5.4.12",              value: BUSINESS_TYPE },
  // businessCategory (2.5.4.15) = type of business
  { name: "businessCategory",      value: BUSINESS_CATEGORY },
  // registeredAddress (2.5.4.26) = city
  { type: "2.5.4.26",             value: CITY },
]);

csr.sign(privateKey, forge.md.sha256.create());

const csrPem = forge.pki.certificationRequestToPem(csr);
const keyPem = forge.pki.privateKeyToPem(privateKey);

// Write files
writeFileSync("zatca-private-key.pem", keyPem);
writeFileSync("zatca-csr.pem", csrPem);

// Base64 DER for API
const csrDer    = forge.asn1.toDer(forge.pki.certificationRequestToAsn1(csr)).getBytes();
const csrBase64 = Buffer.from(csrDer, "binary").toString("base64");

const keyAsn1   = forge.pki.privateKeyToAsn1(privateKey);
const keyDer    = forge.asn1.toDer(keyAsn1).getBytes();
const keyBase64 = Buffer.from(keyDer, "binary").toString("base64");

console.log("\nFiles written: zatca-private-key.pem, zatca-csr.pem\n");

console.log("=== CSR Base64 (paste as csrBase64) ===");
console.log(csrBase64);

console.log("\n=== Private Key Base64 (set as ZATCA_PRIVATE_KEY) ===");
console.log(keyBase64);

console.log("\n=== CSR Subject ===");
console.log(`C=${COUNTRY}, O=${VAT_NUMBER}, OU=${BRANCH_NAME}, CN=${EGS_SERIAL}`);
